### PR TITLE
Review of DM-5002: Make ci_hsc resumable

### PR DESCRIPTION
**Pull request by jdswinbank**
In the case where ci_hsc fails or is aborted, scons should be
resumable. Make this so by changing how function signatures are
created for functools.partial objects by wrapping them in a custom
child class of SCons.Action.FunctionAction.

### CODE REVIEW

File: python/lsst/ci/hsc/__init__.py [hunk 1]
```diff
@@ -1 +1,6 @@
+try:
+    import SCons.Action
+except ImportError:
+    raise ImportError("lsst.ci.hsc cannot be imported outside of a scons script.")
+
```
Comment (jdswinbank) [inline python/lsst/ci/hsc/__init__.py]: This seems like quite an arbitrary restriction: although I don't currently have a use case for importing this outside SCons, I'm not sure why we should disallow it. Can you simply move the use of `FunctionAction` into the `SConstruct` file? `makeScons()` is no longer a class method but a free function, you change the way it's called (`RawValidation.makeScons(REPO, data.dataId)` becomes `makeScons(RawValidation, REPO, data.dataId)`, etc), and this error is no longer necessary.

File: python/lsst/ci/hsc/validate.py [hunk 2]
```diff
@@ -133,7 +143,7 @@ class WarpValidation(Validation):
     _datasets = ["deepCoadd_tempExp", "deep_makeCoaddTempExp_config", "deep_makeCoaddTempExp_metadata"]
 
 class CoaddValidation(Validation):
-    _datasets = ["deepCoadd", "deep_assembleCoadd_config", "deep_assembleCoadd_metadata"]
+    _datasets = ["deepCoadd", "deep_safeClipAssembleCoadd_config", "deep_safeClipAssembleCoadd_metadata"]
```
Comment (jdswinbank) [inline python/lsst/ci/hsc/validate.py]: This change looks like it should be on a separate commit.

File: python/lsst/ci/hsc/validate.py [hunk 3]
```diff
@@ -6,6 +6,16 @@
...
 from lsst.daf.persistence import Butler
+from SCons import Action
+
+class callAction(Action.FunctionAction):
+    def __init__(self, ident, func, **kw):
+        self.ident = ident
+        self.func = func
+        Action.FunctionAction.__init__(self, self.func, kw)
+
+    def get_contents(self, *args, **kwargs):
```
Comment (jdswinbank) [inline python/lsst/ci/hsc/validate.py]: What is `get_contents()` for? You never call it.

File: python/lsst/ci/hsc/validate.py [hunk 4]
```diff
@@ -6,6 +6,16 @@
 import functools
 from lsst.pex.logging import getDefaultLog
 from lsst.daf.persistence import Butler
+from SCons import Action
+
+class callAction(Action.FunctionAction):
```
Comment (jdswinbank) [inline python/lsst/ci/hsc/validate.py]: Commentary describing what's going on here would be helpful. I don't mind if that appears in this class, in the `makeScons` method, or both, but some explanatory text is necessary.

Comment (jdswinbank) [inline python/lsst/ci/hsc/validate.py]: Also, the class name should be capitalized!

File: python/lsst/ci/hsc/validate.py (hunk 3 above)
Comment (natelust) [inline python/lsst/ci/hsc/validate.py]: I should comment this, get_contents is what scons calls to generate a function signature, and should be deterministic. The default get_contents in Action.FunctionAction does not generate consistent signatures across runs of functools.partial, so we overload it to return something deterministic
